src/Codec/LazyBSONArrayCodec.php

```diff
@@ -0,0 +1,110 @@
+<?php
+/*
+ * Copyright 2023-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace MongoDB\Codec;
+
+use MongoDB\BSON\PackedArray;
+use MongoDB\Exception\UnsupportedValueException;
+use MongoDB\Model\LazyBSONArray;
+
+/**
+ * Codec for lazy decoding of BSON PackedArray instances
+ *
+ * @template-implements Codec<PackedArray, LazyBSONArray>
+ */
+final class LazyBSONArrayCodec implements Codec, KnowsCodecLibrary
+{
+    private ?CodecLibrary $library = null;
+
+    public function attachCodecLibrary(CodecLibrary $library): void
+    {
+        $this->library = $library;
```

Comment:
Yes, that is an option and it could be problematic. On the other hand, I expect people to create a codec library containing a number of codecs for types they need (e.g. to encode `UTCDateTime` instances to `DateTimeInterface` instances and back) and then create a `DocumentCodec` with the library they'll need for use with a `Collection` instance. I'll make a note to add examples for this in the documentation.

Reply:
The library could be passed as an argument to `encode`/`decode` methods. So the codec classes are more stateless and the library can be changed for nested objects.

That would also remove the circular dependency `Codec <-> Library`.

Reply:
@alcaeus: I saw @GromNaN LGTM-ed, but it seems like there's still something to be done on this thread. I'll defer to you to resolve.

Reply:
The reason I didn't want to pass this down in `encode` but rather use a library is that a codec will not necessarily want to inherit the library they are a part of. Thinking about this from the Doctrine ODM side, the CodecLibrary will contain a number of common codecs, e.g. those to handle arrays (which will be converted to `PersistentCollection` instances) and other types that require preparation (e.g. the `ObjectId` type which converts the value to a string). A `DocumentCodec` instance for a document will potentially not be part of a codec library and only receive a codec library.

It could be that once I prototype a hydrator based on codecs there will be some more changes to this, so I'm not saying this is here to stay. For example, it could be useful for the lazy BSON structures to be able to select a codec based on a field name, which would also help in the context of the Doctrine ODM. While we only deal with generic data structures (such as the lazy BSON instances) this isn't relevant yet, so I didn't want to engineer something that might not work.

Reply:
SGTM. I'll leave to @GromNaN to resolve after reading.